### PR TITLE
Improved non-HTTP(S) link handling behavior

### LINKED ISSUE #4897: Determine non-HTTP(S) link handling behavior
https://github.com/dotnet/maui/pull/4645 was focused on HTTP(S) link handling. We may want to consider how we're handling non-HTTP(S) links (ex. to open a PDF, app store, etc.).

image

### PR DESCRIPTION
**Pull request by MackinnonBuck**
### Description of Change

In Blazor MAUI for Android, clicking a link with a scheme other than HTTP or HTTPS would previously sometimes crash the app or display a "Webpage not available" page. This PR resolves the undesirable behavior by making the following changes:
* If a link's scheme is not HTTP(S), but the scheme has a registered handler, its associated activity will be started when the link is clicked (this works for schemes for installed apps, the `intent` scheme, etc.).
* If a link's scheme does _not_ have a registered handler, clicking the link does nothing (rather than displaying the "Webpage not available" page). This is consistent with the behavior on other platforms.
* ~~If a link uses the `intent` scheme, but the intent could not be resolved (or the app could not be started) and a `browser_fallback_url` is specified, attempt to start an activity from an intent created by the fallback URL.~~

### Issues Fixed

Fixes #4897

### CODE REVIEW

File: src/BlazorWebView/src/Maui/Android/WebKitWebViewClient.cs
Comment (TanayParikh) [inline src/BlazorWebView/src/Maui/Android/WebKitWebViewClient.cs:140]: Wouldn't we want to reroute the fallback URI to `TryOverrideHttpUrlLoading`?

**Comment by MackinnonBuck** [comment]
Before this gets merged, we will probably want to consider whether the behavior I've described in my previous comment is desirable. The `S.browser_fallback_url` feature is Chrome-specific (not some other standard) so we need to decide if it's worth it to replicate support for it. Removing support for that feature wouldn't infringe on the functionality of "intent"-schemed URIs with registered handlers.

File: src/BlazorWebView/src/Maui/Android/WebKitWebViewClient.cs
Comment (TanayParikh) [inline src/BlazorWebView/src/Maui/Android/WebKitWebViewClient.cs:42]: Is it possible to have a URL with no scheme? Ex. what if the link is `microsoft.com`. Does it get the `http(s)` scheme by default somewhere?

File: src/BlazorWebView/src/Maui/Android/WebKitWebViewClient.cs
Comment (MackinnonBuck) [reply src/BlazorWebView/src/Maui/Android/WebKitWebViewClient.cs:140]: You could have fallback URIs that don't have HTTP(S) schemes. For example, if you're trying to create a link that opens the YouTube app, you might want to open the Google Play store instead using the `market` scheme (e.g. "market://details?id=com.google.android.youtube").

**Comment by TanayParikh** [comment]
I realize this wasn't covered in the original issue, but what happens if you use links like these on other platforms? Specifically, iOS with [custom URL Schemes](https://medium.com/wolox/ios-deep-linking-url-scheme-vs-universal-links-50abd3802f97#:~:text=There%20are%202%20ways%20of%20implementing%20deep%20linking,webpage%20and%20your%20app%20under%20the%20same%20link.)?

It's also possible on Windows (ex. "do you want to allow example.com to open `GitHub`), or deep links to the `Microsoft Store`.

Just understanding if we get crashes on those platforms, similar to what we saw on Android, would be good to understand.

File: src/BlazorWebView/src/Maui/Android/WebKitWebViewClient.cs
Comment (MackinnonBuck) [reply src/BlazorWebView/src/Maui/Android/WebKitWebViewClient.cs:42]: If it's something like `<a href="microsoft.com"></a>`, then the URL scheme is implicitly HTTPS. However, in that specific case, the host is also implicitly 0.0.0.0, which makes the link not work anyway.

**Comment by MackinnonBuck** [comment]
@TanayParikh I've tested links with various schemes on other platforms, and they don't seem to have any issues. If the scheme is known, the link is handled correctly, and if the scheme is unknown, nothing happens when the link is clicked. This is why I changed the Android implementation to no-op when the scheme doesn't have a handler.

File: src/BlazorWebView/src/Maui/Android/WebKitWebViewClient.cs
Comment (MackinnonBuck) [reply src/BlazorWebView/src/Maui/Android/WebKitWebViewClient.cs:140]: Coming back to this, I wonder if there's a scenario where you want the fallback link to be another page in the app? Because if that's the case, we probably would want to at least check if the scheme is `http` in case we want the specialized navigation behavior.

Comment (MackinnonBuck) [reply src/BlazorWebView/src/Maui/Android/WebKitWebViewClient.cs:140]: (And this all depends on whether we think supporting the chrome `S.browser_fallback_url` extra is worthwhile)

File: src/BlazorWebView/src/Maui/Android/WebKitWebViewClient.cs
Comment (SteveSandersonMS) [inline src/BlazorWebView/src/Maui/Android/WebKitWebViewClient.cs:44]: It's a bit odd that we stringify the URL here and then re-parse it in `TryOverrideHttpUrlLoading`. Would it work to keep it in the form of an `Android.Net.Uri` object?

I see it has to be made into a `System.Uri` if we're going to call `ExternalNavigationStarting` but that doesn't happen in most navigations (to internal URLs).

File: src/BlazorWebView/src/Maui/Android/WebKitWebViewClient.cs
Comment (SteveSandersonMS) [reply src/BlazorWebView/src/Maui/Android/WebKitWebViewClient.cs:42]: This makes sense.

Related question: are the supplied `Scheme` values guaranteed to be lowercase all the time? Likewise in the `intent` case below.

File: src/BlazorWebView/src/Maui/Android/WebKitWebViewClient.cs
Comment (SteveSandersonMS) [inline src/BlazorWebView/src/Maui/Android/WebKitWebViewClient.cs:51]: Likewise here. If we passed it as an `Android.Net.Uri`, we wouldn't need the `uriHasIntentScheme` param as the receiving code could encapsulate this check.

File: src/BlazorWebView/src/Maui/Android/WebKitWebViewClient.cs
Comment (SteveSandersonMS) [inline src/BlazorWebView/src/Maui/Android/WebKitWebViewClient.cs:52]: It's a bit subtle that `TryStartActivityFromUri` has that name (`Try...`) yet returns `void`, and that on this branch of the if/else we always return yet on the other we fall through to the base. I'm not suggesting the logic is wrong, just that a year from now, it might be a challenge for us to reason about why it goes like this. The fallthrough structure heightens the risk that we make a change that intends to affect one case but actually affects other cases too.

Would it be viable to do either one or both of the following?

 * Structure the code so there's no fallthrough from one case to another, i.e., that we're in "HTTP URL" or "other scheme URL" or "null URL" mode, and each of those fully encapsulates its own logic and definitely returns (which might involve returning a call to base.ShouldOverrideUrlLoading)
 * Add code comments briefly stating why the choice of whether or not to call base is correct. Perhaps each of the three top-level branches could have a sentence summarizing the reasoning for that case.

Likewise I'm wondering if we should be calling the `ExternalNavigationStarting` event for non-HTTP(S) destinations. My sense is that developers would likely want to have control over such navigations too.

File: src/BlazorWebView/src/Maui/Android/WebKitWebViewClient.cs
Comment (SteveSandersonMS) [inline src/BlazorWebView/src/Maui/Android/WebKitWebViewClient.cs:125]: TBH the sophistication of the logic here is mildly concerning. It will be expensive to keep testing and maintaining this in an ongoing way. We'd need quite a lot of different CTI test flows to exercise all this. Is there much of a chance that people will come up with other sub-cases and ask for other variations in behavior?

Since Google didn't see fit to endow Android WebView with the ability to handle this by default, I wonder if it's really needed for BlazorWebView to do so.

Might it be sufficient to extend the `ExternalNavigationStarting` callback to cover the non-HTTP(S)-URL case too, and then developers can put in whatever logic they want based on the specific URL types they intent to use in their app?

I totally accept I might be missing some important point here. As a non-Android user I have limited intuition for how much intent links are used on that OS and what range of support is needed for a core developer experience. Please let me know if you think I'm failing to grasp how right and important this feature is!

File: src/BlazorWebView/src/Maui/Android/WebKitWebViewClient.cs
Comment (MackinnonBuck) [inline src/BlazorWebView/src/Maui/Android/WebKitWebViewClient.cs:60]: The name `OpenInExternalBrowser` really could mean `OpenInExternalApp`. Should we consider renaming this?

Comment (SteveSandersonMS) [reply src/BlazorWebView/src/Maui/Android/WebKitWebViewClient.cs:57]: Good point. What do you think - `OpenInExternalApp`? `OpenExternal`? `OpenExternally`? I'm fine with any. cc @TanayParikh in case he has opinions since he did the earlier version of this.

File: src/BlazorWebView/src/Maui/Android/WebKitWebViewClient.cs
Comment (SteveSandersonMS) [inline src/BlazorWebView/src/Maui/Android/WebKitWebViewClient.cs:40]: I'm not sure why we're using `IsBaseOfPage` here instead of `IsBaseOf`. Does it matter whether or not it's a page?

I thought that distinction was only relevant when deciding whether to do fallback routing.

**Comment by SteveSandersonMS** [comment]
This is a great improvement - thanks @MackinnonBuck! I think we'll easily be able to understand this logic years from now.

File: src/BlazorWebView/src/Maui/Android/WebKitWebViewClient.cs
Comment (MackinnonBuck) [reply src/BlazorWebView/src/Maui/Android/WebKitWebViewClient.cs:40]: I guess my reasoning was that if we're trying to load a file, it seems like we should give the developer the option to open it in an external app/browser (i.e., we only load URLs in the view without invoking the `ExternalNavigationStarting` callback if they refer to other pages in the app). I could be making a wrong assumption here so I'm certainly happy to change it.

File: src/BlazorWebView/src/Maui/Android/WebKitWebViewClient.cs
Comment (SteveSandersonMS) [reply src/BlazorWebView/src/Maui/Android/WebKitWebViewClient.cs:57]: Ideally the same APIs would have as close-to-equivalent-as-we-can behaviors across all platforms. So if we can do it for iOS too, yes.

> Should we even give an option to open external/internal for intent? What would happen if the user requested OpenInWebview for an intent?

My guess is that developers would only want to use `OpenInExternalBrowser` or `CancelNavigation` for intents. But if they explicitly tell us to `InsecureOpenInWebView`, would it be OK to just let them? I know no good will likely come of it, so they probably shouldn't do it, but if they really want to then I'm not sure it's our responsibility to stop them.

Comment (TanayParikh) [reply src/BlazorWebView/src/Maui/Android/WebKitWebViewClient.cs:57]: > Ideally the same APIs would have as close-to-equivalent-as-we-can behaviors across all platforms. So if we can do it for iOS too, yes.

Assuming we're handling iOS/Windows external schemes in the same relative fashion, I think `OpenExternally` or `OpenExternal` makes sense.

> I know no good will likely come of it, so they probably shouldn't do it, but if they really want to then I'm not sure it's our responsibility to stop them.

Fair enough. Can we just make sure if they _do_ do `OpenInWebView` with an intent it just no-ops instead of crashing (not sure what the current code path will do with intents directly. Once we're in a better state with logging, we may also wish to add some logging in place for this to alert developers of a potentially un-intended code path being executed.

File: src/BlazorWebView/src/Maui/Android/WebKitWebViewClient.cs
Comment (TanayParikh) [reply src/BlazorWebView/src/Maui/Android/WebKitWebViewClient.cs:40]: > would intuitively consider static content in wwwroot to be internal

Do we have webview internal PDF support for instance across all the platforms?

Comment (MackinnonBuck) [reply src/BlazorWebView/src/Maui/Android/WebKitWebViewClient.cs:40]: > (or perhaps routing patterns like /users/{fullname} matching /users/bob.gif - Bob Gif is a possible real name).

The developer could still support this if they wanted by adding their own handler for `ExternalNavigationStarting` and indicating that they want the URL to be opened in the webview (but this does seem like a bit of a hack IMO). Whereas, if we always loaded internal resources in the webview, I don't know that we have a way for the developer to override that behavior to support external navigation.

But there's something else: I think the existing fallback routing logic could still cause problems anyway for URLs ending in `.<something>`. I wonder if we should do something like re-processing the request as a navigation event if we fail to load an internal resource from the request URL (basically falling back to the host page if our initial guess was wrong). That way, users hopefully won't ever get the weird "There was no content at _____" page.

File: src/BlazorWebView/src/Maui/Android/WebKitWebViewClient.cs
Comment (MackinnonBuck) [reply src/BlazorWebView/src/Maui/Android/WebKitWebViewClient.cs:57]: > Would this be specific to android schemes, or would it apply to iOS custom schemes as well? Or is it handled differently for iOS?

iOS will also open links with non-browser apps if a URL scheme is registered with another app (this was something I tested). So even currently, `OpenInExternalBrowser` doesn't always guarantee that the link will be opened in a browser app.

> What would happen if the user requested OpenInWebview for an intent?

IIRC, assuming the scheme is valid, the webview will just display a "Webpage not available" page (but I can double-check this on Monday). Assuming that's the case, I agree that it's probably fine to allow that page to display without enforcing anything ourselves.

**Comment by MackinnonBuck** [comment]
I think the big remaining question here is how we want to handle URLs that look like they point to internal files, but actually refer to pages. Do you have any thoughts on this @SteveSandersonMS @TanayParikh?

My previous comment on this: https://github.com/dotnet/maui/pull/5313#discussion_r830409647

File: src/BlazorWebView/src/Maui/Android/WebKitWebViewClient.cs
Comment (SteveSandersonMS) [reply src/BlazorWebView/src/Maui/Android/WebKitWebViewClient.cs:40]: > But there's something else: I think the existing fallback routing logic could still cause problems anyway for URLs ending in .<something>. I wonder if we should do something like re-processing the request as a navigation event if we fail to load an internal resource from the request URL (basically falling back to the host page if our initial guess was wrong). That way, users hopefully won't ever get the weird "There was no content at _____" page.

Hmm, it's unfortunate that Android's webview doesn't tell us the reason for the URL load event. WebView2 tells us whether or not it's a navigation, so we know unambiguously whether to use fallback logic or whether to regard it as a 404 if there's no match.

On Blazor WebAssembly/Server, we already have a strategy for dealing with not knowing whether it's a navigation: 

 * If a request matches some existing static content, we serve it
 * Otherwise, if it looks like a non-file (has no potential filename extension in the URL) we fall back on the index page
 * Otherwise, 404

It's not a perfect strategy as `/users/bob.gif` won't get the fallback when maybe you want it to (if that's meant to match `/users/{fullname}`). However people can customize this by editing their `Program.cs`. In practice, very few people have ever hit problems with this and complained about it.

> I think the big remaining question here is how we want to handle URLs that look like they point to internal files, but actually refer to pages. Do you have any thoughts on this @SteveSandersonMS @TanayParikh?

(Merging the threads as these perhaps come down to the same question)

I'm definitely happy to give developers control over this. In fact I expect we'll *have to*, because otherwise they have no way to customize this equivalent to what's available for Blazor WebAssembly/Server (i.e., changing the routing config).

The bit I was most thrown off by was referring to loading `/content/file.pdf` as **external** navigation when it's clearly internal to the base href.

Is it possible that the underlying problem here is a matter of terminology? We have this event called `ExternalNavigationStarting`. What if we just passed the whole question on to the app developer by generalizing this to be about all navigations, both internal and external? If we fired an event on *every* URL load, we could preconfigure some `UrlLoadingEventArgs` with a property called `UrlLoadingStrategy` set to whatever our basic guess is (e.g., load internally if an only if it's within the base href - and perhaps in the WebView2 case we take account of whether it's a navigation event since we have that info), then developers can add a handler to customize the logic any way they like for both internal and external loads. Developers who want to override the behavior for a specific subset of URL patterns can match those patterns in their handler, and for all other URLs, leave the preconfigured behavior unchanged.

Does this make sense? Am I up-to-date with all this, or was this already considered before?

Comment (MackinnonBuck) [reply src/BlazorWebView/src/Maui/Android/WebKitWebViewClient.cs:40]: @SteveSandersonMS I'm definitely a fan of the idea of not making a hard distinction in our API between internal and external navigation events. If the rest of the team agrees that it's a good approach, I'm happy to start exploring that route.

One thing to note is that it doesn't seem `ShouldOverrideUrlLoading()` gets invoked for all URLs. I think if it's an internal URL that looks like a page, the method does not get invoked (which I suppose does mean that it is in fact unnecessary to call `AppOriginUri.IsBaseOfPage()`). I've also found that some links with handlers in other apps don't get run through `ShouldOverrideUrlLoading()`, e.g.:
```html
<a href="geo:124.028582,-29.201930" target="_blank">Click here for map</a>
```
The rules around this don't seem to be very clearly [documented](https://developer.android.com/reference/android/webkit/WebViewClient#shouldOverrideUrlLoading(android.webkit.WebView,%20android.webkit.WebResourceRequest)) either.

So, if developers are expecting that they'll be able to modify the way _any_ URL is handled, they could be mistaken. That said, if we made it clear that the event only applies to URLs that could conceivably refer to either a page or a file, then what I've just pointed out is probably a non-issue. We'll just have to make sure that the rules about what URLs can be overridden are consistent enough between each platform for our purposes.

Comment (Eilon) [reply src/BlazorWebView/src/Maui/Android/WebKitWebViewClient.cs:40]: I'm guessing that those WebView events fire if the JS code in the app didn't already handle it. So presumably Blazor's JS code already handled the navigation so there's nothing for the WebView to do.

Comment (MackinnonBuck) [reply src/BlazorWebView/src/Maui/Android/WebKitWebViewClient.cs:40]: > If we're planning on handling this as an "internal" navigation attempt, we'll also have to confirm that PDFs are supported on webviews across all platforms. Not a big deal, but just calling it out.

Do you think it would be fine to just load all "internal" URLs in the webview by default? If the developer can override this behavior, it seems to me like it would be overreaching a bit to decide which file types we should and shouldn't load in the webview by default (but I could be wrong about this).

Comment (TanayParikh) [reply src/BlazorWebView/src/Maui/Android/WebKitWebViewClient.cs:40]: > Do you think it would be fine to just load all "internal" URLs in the webview by default?

Sorry I may have missed something here. Isn't that already our current behavior? Are you suggesting we don't go the route of `UrlLoadingStrategy` and just keep things as is?

>  the developer can override this behavior, it seems to me like it would be overreaching a bit to decide which file types we should and shouldn't load in the webview by default

Normally I'd agree, we want to minimize the complexity however possible. But opening PDFs in the device native PDF viewer seems like a fairly "core" feature. From the user perspective, I much rather have PDFs in the device native app as it's usually a much smoother experience and offers more functionality (like printing, saving to storage, sharing, etc) which may or may not work when it's within the webview. Then again, this may not be something we want to pursue for v1, I'm open to alternatives.

Comment (SteveSandersonMS) [reply src/BlazorWebView/src/Maui/Android/WebKitWebViewClient.cs:40]: @TanayParikh Sorry I didn't see your latest comment before I posted mine. Hence it may sound as if I was ignoring you!

> From the user perspective, I much rather have PDFs in the device native app as it's usually a much smoother experience and offers more functionality (like printing, saving to storage, sharing, etc)

I'm personally ambivalent about whether or not it's good UX to open the PDF in the WebView (one drawback is that it will lose any state held transiently in the hierarchy of Blazor components). I agree it's good that it should be *possible* for this to happen, and for developers to have means of controlling whether or not it does. However it might be too far for us to guarantee cross-platform support for this as it depends on what the platform's native webview happens to do if you point it at a PDF.

Comment (TanayParikh) [reply src/BlazorWebView/src/Maui/Android/WebKitWebViewClient.cs:40]: > Instead, use a button or similar with an @onclick handler that does something useful for the target file (for example, opening it in an external app using MAUI APIs)

Ah I wasn't aware we had MAUI APIs to force the loading into an external (PDF Reader) app. In that case I agree this is all fine given developers at least have the optionality to open externally should it be something they want to do.

Comment (MackinnonBuck) [reply src/BlazorWebView/src/Maui/Android/WebKitWebViewClient.cs:40]: > Sorry I may have missed something here. Isn't that already our current behavior? Are you suggesting we don't go the route of UrlLoadingStrategy and just keep things as is?

I may not have been clear, my bad. I was suggesting that the current default behavior for loading file-like URLs internal to the base href is fine, but we would give developers the option to override it via `UrlLoadingStrategy`.

> Normally I'd agree, we want to minimize the complexity however possible.

I guess my concerns with this were that:

1. If we special-case PDF handling, what should we do about other [image formats](https://en.wikipedia.org/wiki/Comparison_of_web_browsers#Image_format_support) (or other file types that browsers provide a preview for)? Are we going to evaluate their support on each platform and decide what best the default behavior is?
2. Would we ever change the default behavior if a webview on some platform adopted a sufficient built-in PDF viewer? For example, the Android webview doesn't have support for previewing PDFs (so presumably we would open in an external app), but iOS _does_  support it. So should we open within the webview on iOS but not Android? But then what if Android someday gets its own PDF viewer, will we change the default behavior again? Maybe we decide that for now we always open PDFs externally on both platforms, but what if the webview PDF viewers on both platforms improve to the extent where it's unnecessary to do so?

That said, maybe I'm overthinking this and it's fine to only consider the PDF format for now, and always open it externally on mobile platforms. I'm fine taking that route, just wanted to share some potential concerns 🙂

Comment (MackinnonBuck) [reply src/BlazorWebView/src/Maui/Android/WebKitWebViewClient.cs:40]: Oops, I didn't refresh and missed a bunch of comments before posting, sorry!

Comment (MackinnonBuck) [reply src/BlazorWebView/src/Maui/Android/WebKitWebViewClient.cs]: I should be unblocked for now, thanks @SteveSandersonMS! I think most of the work is done, the primary remaining TODOs are:
1. Completing/verifying the new link handling functionality on iOS.
2. Determining what should be done about pages with URLs that look like file names. I've implemented a solution for WinUI, but I'm not sure if we want to include it in this PR (we would probably need something for the other platforms as well).

Hoping to at least get **1.** finished soon so this PR can get reviewed/merged by EOD tomorrow.

**Comment by MackinnonBuck** [comment]
Okay, I think this PR is ready to be reviewed again @SteveSandersonMS @TanayParikh! I'm going to add some thoughts here to provide some context on the changes I made.

***

I changed the `ExternalNavigationStarting` event to `UrlLoading`, so I renamed `ExternalLinkNavigationEventArgs` and `ExternalLinkNavigationPolicy` to `UrlLoadingEventArgs` and `UrlLoadingStrategy` respectively. And, of course, the event is now invoked for internal links in addition to external links (except for those with a `target="_blank"`). The default URL handling behavior follows a simple rule: If the link is internal (i.e. it has an internal host), it gets opened in the webview. Otherwise, it gets opened externally. This behavior can be overridden by changing the `UrlLoadingEventArgs.UrlLoadingStrategy` in the `UrlLoading` event.

This new rule does introduce a subtle change in behavior from an earlier version of this PR. Rather than doing something like `AppOrigin.IsBaseOf(requestUrl)` to determine which links count as "internal", the latest version of this PR simply compares the host names between the app origin URI and the request URI. My reasoning behind this is `IsBaseOf()` will return `false` if the two URIs have different schemes, and I don't think it really makes sense to open an internal links externally if their scheme is not, for example, `https` (or `app` on iOS/Mac Catalyst).

The reason I bring this up is that the default behavior now makes it so if you click a link like `<a href="ftp://0.0.0.0">FTP Link</a>` on Android, then the "Webpage not available" page is displayed in the webview (shown as the motivating example in #4897, the issue that this PR is fixing). While attempting to open the link externally in this specific case results in a no-op rather than an eyesore (assuming there's not an app registered to handle the `ftp` scheme), I still think it's best to default to always opening internal links internally, especially since the behavior can be overridden. Do you think I've reached the correct conclusion here?

***

Okay, on to the next thing: I've renamed `InsecureOpenInWebView` to `OpenInWebView`. It felt weird having the name `Insecure` in something that could be a default behavior, and it's not really insecure if we're loading an internal URL. I think the XML documentation comments for `UrlLoadingEventArgs.UrlLoadingStrategy` and `UrlLoadingStrategy.OpenInWebView` make the risks of opening external links in the webview clear enough, but maybe there's another API design that makes those risks unignorable.

***

And finally, I've also changed the MAUI WinUI fallback routing behavior to support pages with URLs that appear to have file extensions. How this works was outlined in a [previous comment](https://github.com/dotnet/maui/pull/5313#discussion_r834591099). This makes the behavior between all WebView2-based Blazor Hybrid platforms consistent (since WPF and WinForms already behave correctly for pages with "extensions"). I realize this change could be considered slightly off-topic for this PR, so I'm happy to undo the change if we need more time to think over the approach. I think the main downside with this change is that now WinUI is the only MAUI platform that supports this behavior. I haven't been able to come up with a good strategy for iOS/Mac Catalyst and Android yet (e.g. I don't think overzealously falling back on the host page when we get a 404 on an internal resource is desirable).

File: src/BlazorWebView/src/Maui/Windows/WinUIWebViewManager.cs
Comment (SteveSandersonMS) [inline src/BlazorWebView/src/Maui/Windows/WinUIWebViewManager.cs:103]: It's a bit quirky that we're assigning a value to this flag which is never used.

**Comment by SteveSandersonMS** [comment]
Thanks @MackinnonBuck! This has ended up as a fairly significant change but I think it's good we got to this before GA. Letting developers control this behavior for all navigations is a valuable feature.

> Rather than doing something like AppOrigin.IsBaseOf(requestUrl) to determine which links count as "internal", the latest version of this PR simply compares the host names between the app origin URI and the request URI.

I have to admit I don't fully understand the design intent with this. TBH I don't know why anyone would be using the `0.0.0.0` hostname for any practical purpose. Why would someone actually link to `ftp://0.0.0.0/` - what do they want to happen? It seems so strange I'm not sure I'd even mind if the app does crash, unless there's some practical reason I'm missing.

If we could conclude that people aren't actually going to link to `0.0.0.0` on any scheme other than the default one used in the app, then I'd perhaps prioritize keeping the rules simple over worrying about what happens with `ftp://0.0.0.0`. If you're saying that opening internally gives a "page not available" error and opening it externally is a no-op, then both of those sound totally fine to me. I'd more likely go with the more common definition of internal/external (i.e., that the scheme, host, and port must all match, like with `document.origin`) and not try to special-case anything for same-host-different-scheme.

Again, if I'm missing some known use case for same-host-different-scheme (given the host is always `0.0.0.0`), please let me know!

> Okay, on to the next thing: I've renamed InsecureOpenInWebView to OpenInWebView

Sounds good.

> And finally, I've also changed the MAUI WinUI fallback routing behavior to support pages with URLs that appear to have file extensions.

While I agree the behavior you've set up for WinUI is good in isolation, I do wonder if having an inconsistency across WinUI and Android/iOS might end up becoming a problem in itself. Developer might only test on Windows, then be confused about why a link to `/users/bob.gif` correctly loads the "Users" page on Windows but gives a 404 on iOS/Android.

I would probably err on the cautious side by going with a least-common-denominator behavior across all MAUI platforms and filing an issue to consider enhancing this in the future. Is the following viable across all?

 * If the URL matches an existing static asset, use that
 * Otherwise, if we know for sure this is *not* a navigation request (i.e., on WebView2, using the check about `eventArgs.ResourceContext`), then give a 404
 * Otherwise, fall back on the host page

Then routing to `/users/bob.gif` works everywhere (if it's a GIF, we serve it, otherwise we use the host page). In the special case of WinUI and `` then we will give a 404 if no such file exists, whereas on Android/iOS we'd give an HTML response which will lead to a broken image, but that's the developer's fault in creating this error.

**Comment by MackinnonBuck** [comment]
Coming back to my comment here:

> That seems valid to me. I initially thought it might be undesirable to provide a non-404 response for non-navigation requests that also don't match a static asset, but it does seem like that would actually work.

I'm starting to think this might actually be problematic in some cases. While implementing this, I noticed that if, for example, an internal JavaScript file could not be located, you get these strange syntax errors in the console basically complaining that HTML tags are not valid JavaScript, etc. But also, if it's something like `favicon.ico` that isn't validated to the same extent as JavaScript, the 404 errors in the console go away, making the problem much less apparent.

@SteveSandersonMS Do you think these problems are enough to hold off on the navigation fallback behavior changes?

**Comment by TanayParikh** [comment]
cc/ @guardrex we'll have to update https://docs.microsoft.com/en-us/aspnet/core/blazor/hybrid/routing?view=aspnetcore-6.0 based on this new API (after it's finalized/merged). @MackinnonBuck gave a great [summary of the changes in this comment](https://github.com/dotnet/maui/pull/5313#issuecomment-1078511479).

**Review by SteveSandersonMS — APPROVED** [review]
Approving assuming this now reflects the final design that you worked out with @javiercn. Thanks for working through all the details on this one!

File: src/BlazorWebView/src/WindowsForms/BlazorWebView.cs
Comment (TanayParikh) [inline src/BlazorWebView/src/WindowsForms/BlazorWebView.cs:120]:
```suggestion
		[Description("Allows customizing how links are opened. By default, opens internal links in the webview and external links in an external app.")]
```

**Review by TanayParikh — CHANGES_REQUESTED** [review]
LGTM, just a doc change needed 😄

**Comment by TanayParikh** [comment]
image

That timing 😄

File: src/BlazorWebView/src/WindowsForms/BlazorWebView.cs
Comment (TanayParikh) [reply src/BlazorWebView/src/WindowsForms/BlazorWebView.cs:120]: https://github.com/dotnet/maui/pull/5675

**Comment by MackinnonBuck** [comment]
Oh no! Sorry about that @TanayParikh!

**Pull request merged**
